mypy fails importing my package with --silent-imports and -p

When I run mypy over [qutebrowser](https://github.com/The-Compiler/qutebrowser), things look good initially (apart from some use of type annotations for my own purposes):

```
$ mypy -m qutebrowser
qutebrowser/qutebrowser.py:27: note: In module imported here,
qutebrowser.py:24: note: ... from here:
qutebrowser/misc/checkpyver.py:29: error: No library stub file for standard library module 'tkinter'
qutebrowser/misc/checkpyver.py:29: note: (Stub files are from https://github.com/python/typeshed)
qutebrowser/misc/checkpyver.py:33: error: Cannot find module named 'Tkinter'
qutebrowser/misc/checkpyver.py:33: note: (Perhaps setting MYPYPATH or using the "--silent-imports" flag would help)
qutebrowser/misc/checkpyver.py:34: error: Cannot find module named 'tkMessageBox'
qutebrowser/qutebrowser.py:42: note: In module imported here,
qutebrowser.py:24: note: ... from here:
qutebrowser/misc/earlyinit.py:28: error: Cannot find module named 'hunter'
qutebrowser/misc/earlyinit.py:34: error: Cannot find module named 'faulthandler'
qutebrowser/misc/earlyinit.py:40: error: No library stub file for standard library module 'tkinter'
qutebrowser/misc/earlyinit.py:83: error: No library stub file for module 'PyQt5.QtWidgets'
qutebrowser/misc/earlyinit.py:84: error: No library stub file for module 'PyQt5.QtCore'
qutebrowser/misc/earlyinit.py:222: error: No library stub file for module 'PyQt5.QtNetwork'
qutebrowser/qutebrowser.py:156: note: In module imported here,
qutebrowser.py:24: note: ... from here:
qutebrowser/app.py:34: error: No library stub file for module 'PyQt5.QtWidgets'
qutebrowser/app.py:35: error: No library stub file for module 'PyQt5.QtWebKit'
qutebrowser/app.py:36: error: No library stub file for module 'PyQt5.QtGui'
qutebrowser/app.py:37: error: No library stub file for module 'PyQt5.QtCore'
qutebrowser/app.py:40: error: Cannot find module named 'hunter'
qutebrowser/app.py:52: note: In module imported here,
qutebrowser/qutebrowser.py:156: note: ... from here,
qutebrowser.py:24: note: ... from here:
qutebrowser/misc/sessions.py: note: In member "session_save" of class "SessionManager":
qutebrowser/misc/sessions.py:390: error: Parse error before {
qutebrowser/misc/sessions.py:391: error: Parse error before )
qutebrowser/misc/sessions.py: note: In class "SessionManager":
qutebrowser/misc/sessions.py:391: error: Parse error before :
qutebrowser/misc/sessions.py:392: error: Inconsistent indentation
qutebrowser/misc/sessions.py: note: At top level:
qutebrowser/misc/sessions.py:443: error: Inconsistent indentation
```

I then tried using `--silent-imports` as I don't care about the missing stubs for now:

```
$ mypy --silent-imports -m qutebrowser
qutebrowser.py:29: error: "module" has no attribute "main"
```

`--almost-silent` reveals a bit more:

```
$ mypy --silent-imports --almost-silent -m qutebrowser
qutebrowser.py:24: note: Import of 'qutebrowser.qutebrowser' silently ignored
qutebrowser.py:24: note: (Using --silent-imports, module not passed on command line)
qutebrowser.py:24: note: (This note courtesy of --almost-silent)
qutebrowser.py:29: error: "module" has no attribute "main"
```

Note my directory structure looks like this:

```
├── qutebrowser
│   ├── app.py
│   ├── [...]
│   ├── qutebrowser.py
│   ├── resources.py
│   └── utils
│       [...]
├── [...]
├── qutebrowser.py
```

That is, there's a toplevel `qutebrowser.py` which is a simple launcher script:

```python
import qutebrowser.qutebrowser
import sys

if __name__ == '__main__':
    sys.exit(qutebrowser.qutebrowser.main())
```

Then there's the `qutebrowser` package, and a `qutebrowser.py` in it which
starts the actual application (in `main()`).

I should probably rename at least `qutebrowser/qutebrowser.py` and maybe the
toplevel `qutebrowser.py`, but I think there's still no reason mypy should get
confused by it with `--silent-imports`.

`-p` doesn't seem to change anything:

```
$ mypy --silent-imports --almost-silent -p qutebrowser
qutebrowser.py:24: note: Import of 'qutebrowser.qutebrowser' silently ignored
qutebrowser.py:24: note: (Using --silent-imports, module not passed on command line)
qutebrowser.py:24: note: (This note courtesy of --almost-silent)
qutebrowser.py:29: error: "module" has no attribute "main"
```

Using `qutebrowser/` seems to work, i.e. I get some more parse errors which will go away when I fix my annotations :wink:

This whole behaviour strikes me as unintuitive though. Why does `--silent-imports` change the behaviour of `-m` and `-p`? What's the difference between the two? Why couldn't `-m` check the whole package if a package is passed, similarily to what python does? It seems to do exactly that without `--silent-imports`?

--- a/mypy/build.py
+++ b/mypy/build.py
@@ -599,10 +599,14 @@ def find_module(id: str, lib_path: Iterable[str]) -> str:
         sepinit = os.sep + '__init__'
         for base_dir in candidate_base_dirs:
             base_path = base_dir + seplast  # so e.g. '/usr/lib/python3.4/foo/bar/baz'
+            # Prefer package over module, i.e. baz/__init__.py* over baz.py*.
+            for extension in PYTHON_EXTENSIONS:
+                path = base_path + sepinit + extension
+                if is_file(path) and verify_module(id, path):
+                    return path
+            # No package, look for module.
             for extension in PYTHON_EXTENSIONS:
                 path = base_path + extension
-                if not is_file(path):
-                    path = base_path + sepinit + extension
                 if is_file(path) and verify_module(id, path):
                     return path
         return None

--- a/mypy/test/data/check-modules.test
+++ b/mypy/test/data/check-modules.test
@@ -769,3 +769,19 @@ class A:
 [file d.py]
 import a
 class D(a.C): pass
+
+[case testPreferPackageOverFile]
+import a
+[file a.py]
+/  # intentional syntax error -- this file shouldn't be parsed
+[file a/__init__.py]
+pass
+[out]
+
+[case testPreferPackageOverFile2]
+from a import x
+[file a.py]
+/  # intentional syntax error -- this file shouldn't be parsed
+[file a/__init__.py]
+x = 0
+[out]

